### PR TITLE
Remove-unused-identifiers codefix skips assigned identifiers

### LINKED ISSUE #32196: Delete unused declarations codefix should offer "Delete unused elements of only this declaration" too
Issue Type: <b>Bug</b>

When I click 'Delete all unused declarations', my function with three parameters, the second of which was not used, was removed. This is a serious error, it should not be removed, because it makes my logic wrong.

VS Code version: Code 1.35.1 (c7d83e57cd18f18026a8162d042843bda1bcf21f, 2019-06-12T14:30:02.622Z)
OS version: Windows_NT x64 10.0.17763

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i5-8600K CPU @ 3.60GHz (6 x 3600)|
|GPU Status|2d_canvas: enabled<br>checker_imaging: disabled_off<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>native_gpu_memory_buffers: disabled_software<br>rasterization: enabled<br>surface_synchronization: enabled_on<br>video_decode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|23.86GB (17.18GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (33)</summary>

Extension|Author (truncated)|Version
---|---|---
rainbow-brackets|2gu|0.0.6
vscode-file-peek|abi|1.0.1
vscode-browser-preview|auc|0.5.9
path-intellisense|chr|1.4.2
angular-schematics|cyr|1.23.0
vscode-eslint|dba|1.9.0
githistory|don|0.4.6
gitlens|eam|9.8.2
EditorConfig|Edi|0.13.0
tslint|eg2|1.0.43
vscode-npm-script|eg2|0.3.7
tortoise-svn|fan|0.1.1
auto-rename-tag|for|0.1.0
align-mode|Gru|0.0.16
svgeditor|hen|2.9.0
git-graph|mhu|1.9.0
code-beautifier|mic|2.2.1
Angular-BeastCode|Mik|8.0.10
vscode-postgresql|ms-|0.2.0
powershell|ms-|2019.5.0
debugger-for-chrome|msj|4.11.6
sqltools|mtx|0.19.6
angular-console|nrw|8.0.5
material-icon-theme|PKi|3.8.0
vscode-yaml|red|0.4.1
angular-follow-selector|san|1.2.0
open-in-browser|tec|2.0.0
pdf|tom|0.5.0
non-breaking-space-highlighter|vik|0.0.3
vscodeintellicode|Vis|1.1.7
vscode-icons|vsc|8.8.0
vscode-todo-highlight|way|1.0.4
html-css-class-completion|Zig|1.19.0

(1 theme extensions excluded)

</details>

### PR DESCRIPTION
**Pull request by sandersn**
Previously, fixUnusedIdentifier would delete any variable-like declaration with no reads, even if it had writes:

```ts
function f(a) { }
f(1)
let x = 0
x = 1
export {}
```

becomes

```ts
function f() { }
f()
export {}
```

But any kind of argument will be deleted, even one with side effects.

Now fix-all will only delete parameters and variable declarations with no assignments, and a single fix will only delete parameters with no assignments:

```ts
function f(a) { }
function g(a) { }
f(1)
g
let x = 1
let y
```

becomes

```
function f(a) { }
function g() { }
f(1)
g
let x = 1
```

Fixes #32196 ([part 2](https://github.com/microsoft/TypeScript/issues/32196#issuecomment-513279210))

### CODE REVIEW

**Comment by sandersn** [comment]
@amcasey I tagged you for a review of the behaviour since I believe you've thought about this before. @gabritto you might have opinions as well.

File: tests/cases/fourslash/codeFixUnusedIdentifier_parameter_all.ts [hunk 1]
```diff
@@ -5,37 +5,53 @@
 
 ////function f(a, b, { x, y }) { b; }
 ////f(0, 1, { x: 1, y: 1 });
+////function g(a, b, { x, y }) { b; }
+////g;
```
Comment (amcasey) [inline tests/cases/fourslash/codeFixUnusedIdentifier_parameter_all.ts:9]: Would a function with optional parameters be a more realistic example?

**Comment by amcasey** [comment]
```ts
function g(x: number, y: number): void {}
```

Should remove-all remove both?  It seems like non-trailing parameters can be removed if all following parameters are unused.

**Comment by amcasey** [comment]
More generally, do we expect that there will be no more unused declaration suggestions after applying the code fix or might the user need to locate another one and fix-all again?

**Comment by amcasey** [comment]
Before: `function g({ x, y }: { x: number; y: number; }): void {}`
After: `function g({ }: { x: number; y: number; }): void {}`

**Comment by amcasey** [comment]
Before: `function g(x = sideEffects() ): void {}`
After: `function g( ): void {}`

**Comment by amcasey** [comment]
Before:
```ts
function f() {
    const { x, y } = { x: sideEffects(), y: sideEffects() };
}
```
After:
```ts
function f() {
}
```

**Comment by amcasey** [comment]
Is this case interesting?  Your example in the description leads me to believe we'd rather not remove `x`.
```ts
function g(x) { }
const f = g;
f(1);
```

**Comment by sandersn** [comment]
Replies: 

1. `function g(x: number, y: number): void {}`

should result in `function g(): void {}`. My comments in the bug are misleading; it's fine to delete any unused parameters.

2. 
>  do we expect that there will be no more unused declaration suggestions after applying the code fix or might the user need to locate another one and fix-all again?

There shouldn't be any unused declarations left and one fix-all should be enough.

3. `function g({ }: { x: number; y: number; }): void {}` -- I assume that this is your desired output. I added a type to the test case to show that the type is left behind.

4. `function g(x = sideEffects()): void {}` -- this should be preserved. I'll add a test and fix.

5. `const { x, y } = { x: sideEffects(), y: sideEffects() };` -- same

6. I don't mind missing aliasing too much, although it'd definitely be nice; the type system will catch incorrect, explicitly typed aliases -- although removing parameters usually won't add a type error.

**Comment by amcasey** [comment]
1. I don't think it does remove both (or is offered on `x`).  I could be misremembering.
2. Cool
3. I thought that was undesirable.  Why would I require the argument to have two properties I'm not going to use.  I would have thought the parameter would be removed.
4. Sounds good
5. Sounds good
6. It seems weird to me that `g`'s parameter is unused in this case `f(g)`, but used in this case `f(x => g(x))`

**Comment by sandersn** [comment]
Replies:

3. I'll see about removing the parameter entirely.
6. Hmm. I guess it makes sense to treat an uncalled reference to a function as referring to, at least, its required parameters. Probably its optional ones as well. If we wanted to be precise, it would be (1) only when used as an argument and (2) based on the parameter type, but that's the commonest usage so I don't think it's worth the effort.

**Comment by sandersn** [comment]
I think the latest commit about covers it.

**Comment by amcasey** [comment]
@sandersn Cool, I'll have another look.  I should mention that I'm buddy testing and not actually reviewing the code.

**Review by andrewbranch — APPROVED** [review]

**Comment by andrewbranch** [comment]
@typescript-bot pack this

**Comment by andrewbranch** [comment]
No playground? ☹️

**Comment by amcasey** [comment]
(5) appears to still be removed?

**Comment by amcasey** [comment]
Before: `function g([x, y] = [sideEffects(), sideEffects()]): void {}`
After: `function g([] = [sideEffects(), sideEffects()]): void {}`

That seems plausible (maybe?), but this is weird:

Before: `function g([x, y]: number[]): void {}`
After: `function g([]: number[]): void {}`

**Comment by amcasey** [comment]
Do you expect the parameter and the type parameter to be fixed up in a single step (i.e. by fix-all)?
```ts
function g<T>(t: T): void {}
```

**Comment by amcasey** [comment]
Call site isn't cleaned up:
```ts
class Base {
    constructor(x: number) {} // Remote unused parameter
}

class Derived extends Base {
    constructor(x: number) {
        super(x);
    }
}
```

**Comment by sandersn** [comment]
Updates:

1. (5) should no longer be removed; declarations with initialisers are now skipped, unless they're partially-used destructurings, in which case the unused elements are still removed.
2. Array bindings weren't handled completely. I fixed that.
3. Parameter and type parameter should be fixed in a single step, if the example use of `g` is `g()`; if there's no use of `g`, the whole function is deleted.
4. Still need to look at the class example.

**Comment by sandersn** [comment]
https://github.com/microsoft/TypeScript/issues/41521 tracks the fact that the codefix doesn't understand `super` calls.

**Comment by amcasey** [comment]
```ts
function F() {
    const [x, y] = [1, 2];
}
```
Removing unused variable `x` yields
```ts
function F() {
    const [x, y] = [1, 2];
}
```

Is the leading comma expected?

**Comment by amcasey** [comment]
```ts
function F(y: number) {
    var y: number; // Remove all unused here
}
```

`var y` is not removed - only the parameter.

**Comment by amcasey** [comment]
Before:
```ts
function G() {
    let y = 1; // remove y
    y++;
}  
```
After
```ts
function G() {
    y++; // error
}  
```

**Review by amcasey — APPROVED** [review]
This is clearly going to have a bug tail, but I would say that it basically works.

**Comment by sandersn** [comment]
Filed #41618, #41621 to track the last two. I realised I didn't actually test codefix-all for array binding, and had to fix the code, which I did in the latest commit.

**Pull request merged**